sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp

```diff
@@ -0,0 +1,215 @@
+#ifdef USE_ACC_PTR
+    local_accessor<int, 1> *local_acc,
+#else
+    local_accessor<int, 1> local_acc,
+#endif
+    int groupID, int local_id, int *o) SYCL_ESIMD_FUNCTION {
+#ifdef USE_ACC_PTR
+  ESIMD_CALLEE_nbarrier<Groups, Threads, Size>(*local_acc, groupID, local_id, o);
+#else
+  ESIMD_CALLEE_nbarrier<Groups, Threads, Size>(local_acc, groupID, local_id, o);
+#endif
+}
+
+template <int CaseNum, unsigned Groups, unsigned Threads>
+bool test(queue q) {
+  constexpr unsigned Size = VL * Threads * Groups;
+
+  std::cout << "Case #" << CaseNum << "\n Treads: " << Threads
+            << ", Groups: " << Groups << ", Size: " << Size << "\n";
+
+  static_assert(Threads > 1, "Threads number must be greater than 1");
+  static_assert(Threads % 2 == 0, "Threads number expect to be even");
+  static_assert(Groups > 1, "Threads number must be greater than 1");
+
+  auto dev = q.get_device();
+  auto device_slm_size =
+      dev.template get_info<sycl::info::device::local_mem_size>();
+
+  // The test is going to use Size elements of int type.
+  if (device_slm_size < Size * sizeof(int)) {
+    // Report an error - the test needs a fix.
+    std::cerr << "Error: Test needs more SLM memory than device has"
+              << std::endl;
+    return false;
+  }
+
+  auto *out = malloc_shared<int>(Size, q);
+  for (int i = 0; i < Size; i++) {
+    out[i] = -1;
+  }
+
+  try {
+    // workgroups
+    sycl::range<1> global_range{Size};
+    // threads in each group
+    sycl::range<1> local_range{Size / Groups};
+
+    auto e = q.submit([&](handler &cgh) {
+      auto local_acc = local_accessor<int, 1>(Size, cgh);
+      cgh.parallel_for<KernelID<CaseNum>>(
+          nd_range<1>(global_range, local_range),
+          // This test requires an explicit specification of the subgroup size
+          [=](nd_item<1> item) [[intel::reqd_sub_group_size(VL)]] {
+            sycl::group<1> g = item.get_group();
+            sycl::sub_group sg = item.get_sub_group();
+
+            // Thread's ID in ESIMD context
+            int local_id = sg.get_group_linear_id();
+            int groupID = g.get_group_linear_id();
+
+            // SLM init
+            uint32_t slm_id = item.get_local_id(0);
+            auto local_acc_copy = local_acc;
+            local_acc_copy[slm_id] = -1;
+            item.barrier();
+
+#ifdef USE_ACC_PTR
+            auto local_acc_arg = uniform{&local_acc_copy};
+#else
+            auto local_acc_arg = uniform{local_acc_copy};
+#endif
+
+            invoke_simd(sg, SIMD_CALLEE_nbarrier<Groups, Threads, Size>,
+                        local_acc_arg, uniform{groupID}, uniform{local_id},
+                        uniform{out});
+          });
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free(out, q);
+    return false;
+  }
+
+  bool passed = true;
+  for (int i = 0; i < Size; i++) {
+    int ref = i / VL;
+    if (ref % 2 == 1) // every odd ref is skipped
```

Comment:
is that because the off refs are the producers doing an slm write and not a global buffer write?

Reply:
Added a comment to this part as well.
Since the producers have only even thread IDs and are the only ones doing a write, any odd ref must be skipped.